### PR TITLE
[GSOC] Add validator for feasible space distribution

### LINKED ISSUE #2374: [GSOC] Project 8: Support various `Parameter Distribution` in Katib
## Motivation
Currently, [Katib](https://github.com/kubeflow/katib) is limited to supporting only uniform distribution for integer, float, and categorical hyperparameters. By introducing additional distributions, Katib will become more flexible and powerful in conducting hyperparameter optimization tasks.
A Data Scientist requires Katib to support multiple hyperparameter distributions, such as log-uniform, normal, and log-normal, in addition to the existing uniform distribution. For instance, learning rates often benefit from a log-uniform distribution because small values can significantly impact performance. Similarly, normal distributions are useful for parameters that are expected to vary around a central value.
## Goal
The project aims to enhance Katib's Experiment APIs by introducing support for various hyperparameter distributions beyond the current uniform distribution. 
Currently, the Experiment APIs support `parameterType` and `feasibleSpace`. To provide greater flexibility, we expanded the `feasibleSpace` to include a `Distribution` field, allowing users to specify the type of distribution for each parameter.

This Distribution field is designed to be mutually exclusive, meaning it can be left empty, enabling users to tune hyperparameters without specifying a distribution if desired.

### Preparation
- [x] Converted the GSoC proposal into a KEP: #2334
### Development
- [x] Added `Distribution` field to the `FeasibleSpace` object: #2377 
- [x] Updated gRPC APIs to include the `Distribution` field: #2397
- [x] Modified Suggestion client to handle the new Distribution field: #2403 
- [x] Updated validation webhook to support both ParameterType and Distribution: #2404 
- [x] Implemented logic updates for handling new distributions in `hyperopt`suggestion service: #2412
- [ ] Implemented logic updates for handling new distributions in `optuna`suggestion service
- [ ] Implemented logic updates for handling new distributions in `goptuna`suggestion service
### Documentation and Ongoing Work
- [x] Write e2e tests for `normal` and `lognormal` distributions in `hyperopt`: #2412 
- [ ] Extend the Python SDK to include support for the `Distribution` field.
- [ ] Update Katib documentation to reflect new API changes and usage guidelines.
- [ ] Create examples demonstrating the use of different parameter distributions.

## What I learnt during the project
- One of the key hurdles was integrating multiple new distributions into the Katib Experiment APIs while ensuring backward compatibility. Modifying the `feasibleSpace` to include a `Distribution` field required careful consideration of API design and how different suggestion services, like Optuna and Hyperopt, would handle these changes.

- Effective communication is key. The weekly syncs with @andreyvelich and @tenzen-y were extremely valuable, and having a clear agenda beforehand helped me get my questions answered efficiently and stay on track.

- I learned the importance of end-to-end (E2E) tests. By implementing these tests, I gained valuable insights into how the different components of Katib interact and how critical it is to verify that new features, like the distribution support.

## In the End
I am grateful to Google for the opportunity to work on this project and to the Kubeflow community for their support and guidance. Special thanks to my project mentors, @andreyvelich, @tenzen-y, and @terrytangyuan  for your invaluable insights and encouragement. Your mentorship was pivotal in achieving the project's goals.

I hope the enhancements made will significantly benefit users and advance Katib’s capabilities in hyperparameter optimization. Looking forward to continue to contribute to Kubeflow.

### PR DESCRIPTION
**Pull request by shashank-iitbhu**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:
Fixes #2374 

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing

### CODE REVIEW

**Review by tenzen-y — COMMENTED** [review]
As we discussed in today sync, we need to implement the unit testing.

**Review by tenzen-y — COMMENTED** [review]
@shashank-iitbhu Could you add a valid distribution so that we can verify the valid case?

https://github.com/kubeflow/katib/blob/f4ca52640bdc976c9c8c97f665e04b83ce973a39/pkg/webhook/v1beta1/experiment/validator/validator_test.go#L1384

**Comment by shashank-iitbhu** [comment]
> @shashank-iitbhu Could you add a valid distribution so that we can verify the valid case?
> 
> https://github.com/kubeflow/katib/blob/f4ca52640bdc976c9c8c97f665e04b83ce973a39/pkg/webhook/v1beta1/experiment/validator/validator_test.go#L1384

Added a test case for valid distribution.

File: pkg/webhook/v1beta1/experiment/validator/validator.go [hunk 1]
```diff
@@ -19,12 +19,13 @@ package validator
...
 	"encoding/json"
 	"fmt"
-	"k8s.io/apimachinery/pkg/util/validation/field"
 	"path/filepath"
 	"regexp"
 	"strconv"
 	"strings"
 
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
```
Comment (tenzen-y) [inline pkg/webhook/v1beta1/experiment/validator/validator.go:28]:
```suggestion
```

File: pkg/webhook/v1beta1/experiment/validator/validator_test.go [hunk 2]
```diff
@@ -454,6 +455,30 @@ func TestValidateParameters(t *testing.T) {
...
+		{
+			parameters: func() []experimentsv1beta1.ParameterSpec {
+				ps := newFakeInstance().Spec.Parameters
+				ps[0].FeasibleSpace = experimentsv1beta1.FeasibleSpace{
+					Max:          "5",
+					Min:          "1",
+					Distribution: experimentsv1beta1.DistributionUniform,
+				}
+				return ps
+			}(),
```
Comment (tenzen-y) [inline pkg/webhook/v1beta1/experiment/validator/validator_test.go:478]:
```suggestion
			parameters: newFakeInstance() ,
```

We already set the valid parameters in the `newFakeInstance` function, right?

File: pkg/webhook/v1beta1/experiment/validator/validator_test.go [hunk 3]
```diff
@@ -454,6 +455,30 @@ func TestValidateParameters(t *testing.T) {
...
+			wantErr:         nil,
+			testDescription: "Valid distribution case",
```
Comment (tenzen-y) [inline pkg/webhook/v1beta1/experiment/validator/validator_test.go:480]:
```suggestion
			testDescription: "Valid parameters case",
```

**Comment by tenzen-y** [comment]
@shashank-iitbhu Could you address lint errors?

File: pkg/webhook/v1beta1/experiment/validator/validator.go (hunk 1 above)
Comment (tenzen-y) [inline pkg/webhook/v1beta1/experiment/validator/validator.go:28]: Could you consolidate this dependency group into the Line 29-35 group?
After you remove the line 28, you run `make update`.

Comment (shashank-iitbhu) [reply pkg/webhook/v1beta1/experiment/validator/validator.go]: done

File: pkg/webhook/v1beta1/experiment/validator/validator_test.go (hunk 3 above)
Comment (shashank-iitbhu) [reply pkg/webhook/v1beta1/experiment/validator/validator_test.go]: done

**Review by tenzen-y — COMMENTED** [review]
Thanks!
/lgtm
/approve

**Pull request merged**